### DISCUSSION TITLE
How to Send Device Data From EdgeX to AWS IoT Core using Docker Compose File

### DISCUSSION THREAD

**Discussion by samirlenka** [edgexfoundry/community#249] [Q&A]
I am using Docker compose to write "app-service-mqtt" Application Service but not able to connect to AWS IoT Core getting these below errors -
![image](https://github.com/edgexfoundry/community/assets/57938938/21d01c62-bcd8-4248-991c-63446876b889)

1. Also attaching a Docker Compose File YMl for your reference. I am not sure how to pass the password and Username along with Private , Public and Root Key ?? - 
--------------------------------------------------------------------------------------------------------------

  app-service-mqtt:
    image: edgexfoundry/app-service-configurable:2.0.0
    ports:
      - 127.0.0.1:59789:59789/tcp
    read_only: true
    security_opt:
    - no-new-privileges:true
    user: 2002:2001
    container_name: edgex-app-service-configurable-mqtt
    hostname: edgex-app-service-configurable-mqtt
    networks:
      edgex-network: {}
    env_file:
      - .env
    environment:
      CLIENTS_CORE_COMMAND_HOST: edgex-core-command
      CLIENTS_CORE_DATA_HOST: edgex-core-data
      CLIENTS_CORE_METADATA_HOST: edgex-core-metadata
      CLIENTS_SUPPORT_NOTIFICATIONS_HOST: edgex-support-notifications
      CLIENTS_SUPPORT_SCHEDULER_HOST: edgex-support-scheduler
      DATABASES_PRIMARY_HOST: edgex-redis
      EDGEX_PROFILE: mqtt-export
      SERVICE_HOST: edgex-app-service-configurable-mqtt
      Service_Port: 48097
      MessageBus_SubscribeHost_Host: edgex-core-data
      Binding_PublishTopic: events

      Writable_Pipeline_Functions_MQTTSecretSend_Parameters_brokeraddress:

      Writable_Pipeline_Functions_MQTTSecretSend_Parameters_topic:""
      Writable_Pipeline_Functions_MQTTSecretSend_Parameters_clientid: 

      # Auth with username and password (JWT)
      Writable_Pipeline_Functions_MQTTSecretSend_Parameters_authmode: usernamepassword 

      Writable_Pipeline_Functions_MQTTSecretSend_Parameters_skipverify: "true"

      Writable_InsecureSecrets_mqtt_Secrets_username: ""

      # Paste your JWT. JWT can be generated using create_jwt.py
      Writable_InsecureSecrets_mqtt_Secrets_password: ""

      # Contents from rootca.pem
      Writable_InsecureSecrets_mqtt_Secrets_cacert: |
        -----BEGIN CERTIFICATE-----
        MIIDQTCCAimgAwIBAgITBmyfz5m/jAo54vB4ikPmljZbyjANBgkqhkiG9w0BAQsF

**Comment by lenny-goodell** [comment]
The problem you are having is that you are running edgex in **secure mode**, but try to set you secrets using `InsecureSecrets`, `InsecureSecrets` are only used in **non-secure mode**.

You need to run EdgeX in **non-secure mode** using the `no-secty` option when you generate your compose file.
`make gen no-secty asc-mqtt`

See the compose builder README for complete details:
https://github.com/edgexfoundry/edgex-compose/blob/napa/compose-builder/README.md#makefile

Also, the version 2.0.0 that you are using is no longer supported. The latest and only supported version of EdgeX is 3.1.0 (Napa)

**Comment by samirlenka** [comment]
Thank you for your response. I tried with Napa but getting this below error - Kindly help

![image](https://github.com/edgexfoundry/community/assets/57938938/520095c8-e17d-4ec5-a6ea-ddb908cd5ace)

**Reply by lenny-goodell** [reply]
Please check log for Core Metadata to see why it isn't running.

**Reply by samirlenka** [reply]
Looks like Core metadata is running fine attaching the screenshot below for your reference - 
![image](https://github.com/edgexfoundry/community/assets/57938938/d9226b5a-5690-4d2e-91bf-daa7223b51e3)

But looks like issue with Rules engine - 
![image](https://github.com/edgexfoundry/community/assets/57938938/48826326-f600-4c39-8f26-f57fecb7e308)

All these services in the UI also not running - 

![image](https://github.com/edgexfoundry/community/assets/57938938/d3cb7c76-d0c3-4d68-9713-f04fb2771615)
![image](https://github.com/edgexfoundry/community/assets/57938938/2db4f5a2-3a2b-426c-a04e-dbab8c35a0c1)
![image](https://github.com/edgexfoundry/community/assets/57938938/05a7f46f-e57d-45bc-990d-9f0435d0dd7a)

**Reply by lenny-goodell** [reply]
This issue is with Core Metadata not responding to requests. The reason why is that it is listening on `localhost` which doesn't work when running in Docker.  It should be using `edgex-core-metadata` as the host, which is the docker network name for it's container.

Try do a clean restart using `make clean` and then `make run no-secty` from edgex-compose.

**Reply by samirlenka** [reply]
Thank you for your response, now the Docker Compose file is running fine - 

But I am getting stuck in Sending data from EdgeX to AWS IoT Core, and I am not sure where to set Private public and rook keys to connect to AWS IoT Core Thing also not sure where to diclare Client ID and AWS Endpoint. Could you please help with this.

The Compose file which I have generated has these details - 
------------------------------------------------------------------
app-mqtt-export:
    container_name: edgex-app-mqtt-export
    depends_on:
      consul:
        condition: service_started
        required: true
      core-common-config-bootstrapper:
        condition: service_started
        required: true
      core-metadata:
        condition: service_started
        required: true
    environment:
      EDGEX_PROFILE: mqtt-export
      EDGEX_SECURITY_SECRET_STORE: "false"
      SERVICE_HOST: edgex-app-mqtt-export
      WRITABLE_LOGLEVEL: INFO
      WRITABLE_PIPELINE_FUNCTIONS_MQTTEXPORT_PARAMETERS_BROKERADDRESS: AWS Endpoint???
      WRITABLE_PIPELINE_FUNCTIONS_MQTTEXPORT_PARAMETERS_TOPIC: edgex-events
    hostname: edgex-app-mqtt-export
    image: nexus3.edgexfoundry.org:10004/app-service-configurable:latest
    networks:
      edgex-network: null
    ports:
      - mode: ingress
        host_ip: 127.0.0.1
        target: 59703
        published: "59703"
        protocol: tcp
    read_only: true
    restart: always
    security_opt:
      - no-new-privileges:true
    user: 2002:2001
    volumes:
      - type: bind
        source: /etc/localtime
        target: /etc/localtime
        read_only: true
        bind:
          create_host_path: true

**Reply by lenny-goodell** [reply]
@samirlenka , the key can go into InsecureSecrest and set the AuthMode to `clientcert`
See https://docs.edgexfoundry.org/3.2/microservices/application/services/AppServiceConfigurable/details/AvailablePipelineFunctions/#mqttexport for details
These can be set via `WRITABLE_INSECURESECRETS_MQTT_SECRETDATA_CLIENTCERT` and `WRITABLE_INSECURESECRETS_MQTT_SECRETDATA_CLIENTKEY`

The clientId is set via `WRITABLE_PIPELINE_FUNCTIONS_MQTTEXPORT_PARAMETERS_CLIENTID` and the AWS endpoit is set via `WRITABLE_PIPELINE_FUNCTIONS_MQTTEXPORT_PARAMETERS_BROKERADDRESS`
You will also need to set  `WRITABLE_PIPELINE_FUNCTIONS_MQTTEXPORT_PARAMETERS_TOPIC` appropriatly.
